### DISCUSSION TITLE
*UPDATED* Now available: v4 previews!

### DISCUSSION THREAD

**Discussion by vnbaaij** [microsoft/fluentui-blazor#881] [Announcements]
As mentioned already at the end of the [3.0 announcement blog post](https://baaijte.net/blog/microsoft-fast-components-fluentui-3.0.0-ga/) (and repeated here), we are making some big changes for the V4 version of the library:

1) We are changing the root namespace from `Microsoft.Fast.Components.FluentUI` to `Microsoft.FluentUI.AspNetCore.Components`
This is part of the process of becoming independent from the FAST team and getting closer aligned to the ASP.NET Core Blazor team. The `Microsoft` and `FluentUI` parts speak for themselves, we think. By adopting the `AspNetCore.Components` part we will be in line with the standard ASP.NET Core Blazor component naming scheme. 
By choosing this namespace scheme, we also leave room for other possible future Fluent UI implementations that might get distributed by means of NuGet Packages. Think for example about implementations for WPF, WinForms, Avalonia, Uno, etc. **We do not have knowledge of any plans/development/upcoming releases of any of these**
From a code perspective this means you will need to update your `_Imports.razor` and change all `@using...` and `using...` statements in your `.razor` and `.cs` files in your projects.
2) For the v4 version, we will **only support .NET 8 and higher**. We made this choice to be able to fully support new functionality and capabilities that have been added to Blazor since .NET 6. Some of the things we are already changing: support for @bind:after, support the Blazor rendermodes, support Sections and many more. 

If you are staying on .NET 6 or 7, the v3 version of the library will remain available and supported as long as those versions of .NET are supported. Most probably though, we will not be adding any new functionality to the v3 version anymore.  **This means we will remove the .NET 8 support in the v3 version in a future update**

In other words:
- You target .NET 6 or 7 --> Use the v3 packager
- You target .NET 8 --> Use the v4 packages

To help you get started with migrating to this new version, we uploaded the preview packages to NuGet: 
- https://www.nuget.org/packages?q=Microsoft.FluentUI.AspNetCore

Current status:
## V4.0.0-preview.1
- Change root namespace to `Microsoft.FluentUI.AspNetCore.Components`

## V4.0.0-preview.2
- Functionally equivalent to v3.2.2
- Replace `AfterBindValue` with native `@bind-Value:after`
- Fix missed or incorrectly replaced namespace errors 
- Update Fluent UI System icons to 1.1.221
 
﻿## V4.0.0-preview.3
- Fix [#891](https://github.com/microsoft/fluentui-blazor/issues/891): JSDisconnectedException in FluentOverflow
- Fix [#861](https://github.com/microsoft/fluentui-blazor/issues/861): Needed extra `empty-content-cell` class 
- FluentCard: Add Width and Height + docs/examples
- FluentToast: Timeout is now in milliseconds
- FluentToastContainer renamed to FluentToastProvider
- FluentMessageBarContainer renamed to FluentMessageBarProvider

You can also follow along with the progress we are making in the [demo and documentation site for this version](https://black-pebble-0dc79cb03-879.westeurope.3.azurestaticapps.net/)

**Comment by Narvalex** [comment]
Will you still be based on the [web components develped by the Fast team](https://web-components.fluentui.dev/?path=/docs/concepts-introduction--page)? Or will be you more independent like the [Fluent UI 9 project in React](https://react.fluentui.dev/?path=/docs/concepts-introduction--page)?

**Reply by vnbaaij** [reply]
Yes, **no changes** there. 

The link you posted points to the next version (v3) of the web-components which will fully support the Fluent 2 design. For v4 of our library we will still be on the web components v2 release (which supports the Fluent 1 design).  
Fully supporting the v3 web components is on our roadmap and we are working with the FAST team to achieve this. As I see it now, this will probably be a v5 release of the library.

We also have a number of components  in the library now which are not based on web components. Where possible, we already try to implement the Fluent 2 design for those components. We've done this for example for the MessageBar and Tooltip. We want to keep refining components when possible.

As a sidenote, there is work going on to also base the Fluent UI React components on the new v3 web components. I'm not so familiar with that project so don't ask me for timelines on that or expected versions where this will happen in.

**Comment by mahald** [comment]
But should the naming then not be Microsoft.AspNetCore.Components.FluentUI ?:

![image](https://github.com/microsoft/fluentui-blazor/assets/6714894/e46d8951-3c65-40a7-a1e2-7338dda7929a)

![image](https://github.com/microsoft/fluentui-blazor/assets/6714894/408c6a11-d9ef-49e8-aac2-abf2f32a49a0)

**Reply by vnbaaij** [reply]
It could be **if** we were part of the actual Blazor team, but we're not. We are however closely aligned so that is why we decided (together with them) to go with this naming scheme.

**Comment by horaciodiez** [comment]
Is there a chance we have a "draggable" Dialog in this version?  Thank you!

**Reply by vnbaaij** [reply]
The usage guidelines for the Dialog React component on the Fluent 2 site (https://fluent2.microsoft.design/components/web/react/dialog/usage) does not describe this as a behavior it should expose, so probably not.

**Reply by horaciodiez** [reply]
Thank you for your kind and clear answer.
Ouch! Same as ContentDialog in WinUI. A real pitty. LOB apps not only need draggable but also nested (forbidden! forbidden!) dialogs.
Thank you anyway!

**Reply by nikolayvpavlov** [reply]
Beg to differ. I am yet to see a corporate user dragging anything but a file / mail from Explorer / Outlook into his app :-)

**Comment by dodyg** [comment]
> You can also follow along with the progress we are making in the [demo and documentation site for this version](https://github.com/microsoft/fluentui-blazor/discussions/881)

This points to the wrong URL

**Reply by vnbaaij** [reply]
Oops. Fixed!

**Reply by LockTar** [reply]
@vnbaaij stil a 404. Can you update the correct url? Thanks

**Reply by vnbaaij** [reply]
It has changed again already 😂. It is at https://preview.fluentui-blazor.net now and will be running at https://www.fluentui-blazor.net pretty soon (i.e. rc 1 release imminent 😉)

**Comment by LockTar** [comment]
@vnbaaij I installed the v4 project templates and I see the following in the project file:

```xml
  <PropertyGroup>
    <Version>4.0.0-preview.1</Version>
    <AssemblyVersion>3.2.1.23296</AssemblyVersion>
    <FileVersion>3.2.1.23296</FileVersion>
    <InformationalVersion>3.2.1.23296.1</InformationalVersion>
  </PropertyGroup>
``` 

For what is that? I'm using preview 2 but see preview 1. I also see that AssemblyVersion etc are pointing to v3.2.1? Is there some docs about this PropertyGroup? What does it do? Thanks

**Reply by vnbaaij** [reply]
No, there is no documentation fot that. It is part of our internal build process magic 😉. The only version you will see in the project files moving forward, is the version that we use for the foorer of the demo site. versioning of the files, assembly version and packages will be done in our (internal) Azure DevOps pipeline. This is internal because we need to sign the packages before we can upload them to NuGet. 

The Templates package that is out now does not have all the changes needed yet. I am now working on an rc 1 release that will be more complete. Will report back here once those packages are on NuGet.
